---
title: "Getting Data and Databases - HW"
author: "Chris deRoux"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, error = TRUE, warning = FALSE)
```

---

## Assignment 6

#### <ins> is due on **March 8th, 2024** <ins>

---

**Instructions:** 

Please turn in the answers to this assignment as a knitted R markdown document. 

- For this, don't forget to comment out all package installations after you install them! You don't want to keep installing things over and over!

Answers will be graded for correctness, completeness, and how well the instructions are followed. 

Turn in your assignment on canvas

```  
@ These are the learning objectives associated with each question
```

---

#### QUESTION 1 (5 pts) 

```
@ Students will display that they understand how to obtain data from the SRA.
```

Go to the [SRA website](https://www.ncbi.nlm.nih.gov/sra).

This time, search for an organism of your interest, along with a tissue or condition of your interest. You can further narrow it down to a type of NGS platform (i.e., RNAseq) if you'd like. 

Find something that you are interested in and that seems like a good candidate. Don't get anything too big either. 

In your search results, **click on your sample of interest**. You should be pulled to a new page. Here, you will see a bunch of information on your sample. Under **Runs**, go ahead and click on your sample accession number. This will pull you to the SRA browser were you can see **Metadata, Analaysis, Reads, Data Access, and FASTA/FASTQ download tabs.** 

Go under FASTA/FASTQ and download the FASTQ file.

Go ahead and put your FASTQ files in a new directory called seq_data. Make sure all file paths are correct!

Now read your file in. 
```{r}
#if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

#BiocManager::install("ShortRead")
library(ShortRead)

# this is just an example of what I did. You will have to modify the path for your specific data. 
#seq1 <- readFastq("./seq_data/SRR27190540.fastq.gz", pattern=character(0))
seq1 <- readFastq("C:/Users/Chris/Documents/CM515-course-2024/modules/06_Getting_Data/SRR27781080.fastq(1).gz", pattern=character(0))

# returns sequence information 
sread(seq1) 

# returns the Phred quality score 
quality(seq1) # Phred-scaled quality score represents how confident we are in the assignment of each base call by the sequencer.

# returns the read ID and length 
id(seq1) 
```

go ahead and read in another sample file or two from the same BioProject. You can add your additional code below. 
```{r}
# again these file names are just examples of what i did, you will have to modify the paths. 

seq2 <- readFastq("C:/Users/Chris/Documents/CM515-course-2024/modules/06_Getting_Data/SRR27781081.fastq.gz", pattern=character(0))
```

---

#### QUESTION 2 (5 pts) 

```
@ Students will do some basic analysis on their downloaded FASTQ file(s)
```

Using the samples you read in above, lets try out some more of shortread's functions to interrogate the data. 

```{r}
head(seq1)
head(seq2)
```

```{r}
summary(seq1)
summary(seq2)
```

```{r}
# Performs qualtiy assessment no short reads

qa_data <- qa(seq1, lane=character(1), type="fastq")
# this will output some information about your data
qa_data2 <- qa(seq2, lane=character(1), type="fastq")
c(qa_data, qa_data2)

```

Are there any interesting things about your data files that you see?

```
put your comments here:
Doing some quick back of the envelope calculation, the GC content of the root sample transcripts is 44.7%, while that of the leaf sample transcripts is 45.1%. This is surprisingly low for sequences which are by definition all coding sequences. This could indicate that AT bias is a much stronger ratchet than I thought, or perhaps flanking sequences like poly A tails are being sequenced by Trinity. 
The read quality score for the root samples appears lower than those of the leaf samples, while the read density is similar. I work with an assembled transcriptome of this species, so going forward I will keep in mind that differential expression between root and leaf associated genes may be difficult to interpret granted the differences in raw read quality. I also must wonder whether root transcriptomes are intriniscally problematic due to the abundance of stymbiotic organisms associated with the roots of all plants but the roots of hemiparasitic Epipogium in specific. Hmm. 

```

We are pretty limited by what we can do not together and within the scope of the class, so we wont delve deeper into the other shortread functions. But go ahead and check them out.  

There are other packages like shortread including "Rqc".

[Rqc](https://www.rdocumentation.org/packages/Rqc/versions/1.6.2) is an optimized tool designed for quality control and assessment of high-throughput sequencing data. It performs parallel processing of entire files and produces a report which contains a set of high-resolution graphics.


---

#### QUESTION 3 (5 pts) 

```
@ Students will gain an understanding of the different types of databases out there, how they differ, and what they offer. 
```

Pick one **primary database** that you think could be relevant to your research or interests and describe what data you could obtain from there that would help you moving forward. 

Now do the same with one **secondary database** type.

Finally, do the same for one **specialized database** type.

```
Place your responses here:
I use Genbank regularly to download assembled annotated genomes and SRAs in case assemblies and annotations are unavailable. This data is core to my project, as the short duration of my degree does not allow for de novo sequencing and assembly of plant genomes, unfortunately. 

Now that I know about it, Interpro will be key to my research going forward. My project involves studying evolutionary rate covariation between plant genes in the same genome and between genes in different genomes. The point is to characterize functional differences between parasitic plants and their nearest autotrophic relatives. Seeing as parasites are universally poorly characterized, I will be relying on Gene Ontology for functional characterizations of ERC hits, alongside comparisons to the model plant Arabidopsis whose genome is thoroughly annotated. With a database like Interpro, I can compare strong ERC signals of unknown function to similar protein sequences/motifs and go some way towards characterizing my hits. 

Even more specifically than that, KEGG's PATHWAY database would fit my project perfectly. Seeing as the transition from autotroph to parasite involves a wholesale rearrangement of the organism's metabolism, I expect many of my ERC hits to be involved in metabolic pathways. I expect my data to represent a crude interactome of the functional products of various protein coding genes, so having this database available with so much information on metabolic pathways and other systems should prove very helpful. 

```

---

#### QUESTION 4 (5 pts) 

```
@ Students will learn about a specific R function of their choice, and impliment an example of its usage. 
```

Choose an R function you think is interesting and would like to learn more about. Also mention the R package it can be found in. 

Describe what the function does or how it can be used. 

Give a functional example of its use. I want to try it too. 

I chose two functions which trim low quality tails from raw reads, as a basic step in SRA processing. I've never tried this in R so this is a first for me. trimTailw uses a sliding window frame to decide where to trim read tails, while trimTails tallies up nucleotides falling below a specified quality and trims once enough poor quality nucleotides accumulate. 

I define a few variables here to store the trimmed ShortReadQ objects. It looks like, granted the parameters I set, trimTailw pruned more data from the SRA than trimTails. Without further investigation into the kind of trimming done, I'm not sure which method is more appropriate. It's nice to see that R can do this on its own, with fairly minimal impact on my local system memory given the size of my SRAs. Both methods take a trivial amount of time. 

```{r}

seq1TrimW <- trimTailw(seq1, 1, 1, halfwidth = 5)

print(seq1TrimW)

seq1TrimS <- trimTails(seq1, 1, 1)

print(seq1TrimS)

```

Also provide a link to some documentation describing this function. The **[NAME](url)** notation is used for creating a hyperlink. You can provide the link below by filling in the syntax: 

**[Bioconductor](https://bioconductor.org/packages/release/bioc/html/ShortRead.html)**
